### PR TITLE
eve::map and callable behavior rework

### PR DESCRIPTION
**Pull request by SadiinsoSnowfall**
@DenisYaroshevskiy let me know what you think

### CODE REVIEW

File: include/eve/arch/cpu/wide.hpp [hunk 1]
```diff
@@ -70,6 +70,14 @@ namespace eve
   {
     using storage_base = detail::wide_storage<as_register_t<Type, Cardinal, abi_t<Type, Cardinal>>>;
 
+    // split_type does not exist on cardinals of size 1
+    // this function returns a void_t if the split_type does not exist
+    template <typename T, typename N>
+    static constexpr auto get_split_type() noexcept
+    {
+      if constexpr (N::value > 1) return wide<T, typename N::split_type>{};
+    }
```
Comment (DenisYaroshevskiy) [inline include/eve/arch/cpu/wide.hpp:79]: Why is this function needed? I am not wild about it in the public api.
Also - it's just void, not void_t

File: include/eve/arch/cpu/logical_wide.hpp [hunk 2]
```diff
@@ -78,6 +78,12 @@ namespace eve
     //! Type representing the numerical value associated to the mask
     using mask_type = wide<Type, Cardinal>;
 
+    //! Type representing a logical wide of the same type but with a cardinal twice the size
+    using combined_type = logical<wide<Type, typename Cardinal::combined_type>>;
+
+    //! Type representing a logical wide of the same type but with a cardinal half the size
+    using split_type    = logical<typename wide<Type, Cardinal>::split_type>;
+
```
Comment (DenisYaroshevskiy) [inline include/eve/arch/cpu/logical_wide.hpp:86]: I'm not sure I like these.

File: include/eve/detail/apply_over.hpp [hunk 3]
```diff
@@ -18,25 +18,32 @@ namespace eve::detail
...
     constexpr bool any_emulated   = has_emulated_abi_v<T0>   || (has_emulated_abi_v<T>   || ...);
-    if constexpr(any_aggregated) return aggregate(f, arg0, args...);
-    else if constexpr(any_emulated)   return map(f, arg0, args...);
-    else                              return f(arg0, args...);
 
+    if      constexpr(any_aggregated) return aggregate(f, arg0, args...);
+    else if constexpr(any_emulated)
+    {
+      using r_e = decltype(f(std::declval<element_type_t<T0>>(), std::declval<element_type_t<T>>()...));
+      using r_t = as_wide_t<r_e, fixed<std::max({cardinal_t<T0>{}, cardinal_t<T>{}...})>>;
```
Comment (DenisYaroshevskiy) [inline include/eve/detail/apply_over.hpp:32]: Some requirements/code mismatch.

If `requires simd_value<T0>` - then we know for a fact that all values better have cardinal_v<T0>. If it can be scalar - then the requirement is incorrect.

File: include/eve/detail/apply_over.hpp [hunk 4]
```diff
@@ -18,25 +18,32 @@ namespace eve::detail
...
-    if constexpr(has_aggregated_abi_v<T>)     return aggregate(f, v);
-    else if constexpr(has_emulated_abi_v<T>)  return map(f, v);
-    else                                      return f(v);
+    if      constexpr(has_aggregated_abi_v<T>) return aggregate(f, v);
+    else if constexpr(has_emulated_abi_v<T>)
+    {
+      using r_t = as_wide_as_t<decltype(f(std::declval<element_type_t<T>>())), T>;
+      return map(as<r_t>{}, f, v);
+    }
+    else return f(v);
```
Comment (DenisYaroshevskiy) [inline include/eve/detail/apply_over.hpp:47]: Can you check if we can delete this overload? I can't imagine why we need it.

File: include/eve/detail/function/simd/arm/neon/friends.hpp [hunk 5]
```diff
@@ -39,7 +39,7 @@ namespace eve::detail
     else  if constexpr(c == category::uint8x8 ) return vmvn_u8(v);
     else  if constexpr(c == category::int8x16 ) return vmvnq_s8(v);
     else  if constexpr(c == category::uint8x16) return vmvnq_u8(v);
-    else                                        return bit_cast(~bit_cast(v,as<i_t>{}), as(v));
+    else                                        return bit_cast(~bit_cast(v,as<i_t>{}), as{v});
```
Comment (DenisYaroshevskiy) [inline include/eve/detail/function/simd/arm/neon/friends.hpp:42]: why this change

File: include/eve/module/bessel/regular/cyl_bessel_i0.hpp [hunk 6]
```diff
@@ -16,9 +16,11 @@ namespace eve
...
   struct cyl_bessel_i0_t : elementwise_callable<cyl_bessel_i0_t, Options>
   {
-    template<eve::floating_value T>
-    EVE_FORCEINLINE constexpr
-    T operator()(T a) const noexcept { return EVE_DISPATCH_CALL(a); }
+    template<floating_value T>
+    EVE_FORCEINLINE constexpr T operator()(T a) const noexcept
+    {
+      return this->behavior(as<T>{}, eve::current_api, this->options(), a);
+    }
```
Comment (DenisYaroshevskiy) [inline include/eve/module/bessel/regular/cyl_bessel_i0.hpp:23]: What do these changes have to do with map?

File: include/eve/detail/function/simd/arm/neon/friends.hpp [hunk 7]
```diff
@@ -126,7 +126,7 @@ namespace eve::detail
       else  if constexpr( cat == category::uint64x2)  return vcltq_u64(v, w);
     }
     else  if constexpr( sizeof(T) == 8 )
-      return map([]<typename E>(E const& e, E const& f){ return as_logical_t<E>(e < f); }, v, w);
+      return map(as<logical<wide<T, N>>>{}, []<typename E>(E const& e, E const& f){ return as_logical_t<E>(e < f); }, v, w);
```
Comment (DenisYaroshevskiy) [inline include/eve/detail/function/simd/arm/neon/friends.hpp:129]: why is map requiring an explicit return type good?

**Review by DenisYaroshevskiy — COMMENTED** [review]
I'm a bit confused to what's happening. FOr some reason a lot of changes to macros mixed in with some map stuff. Why?

File: include/eve/module/bessel/regular/cyl_bessel_i0.hpp (hunk 6 above)
Comment (DenisYaroshevskiy) [reply include/eve/module/bessel/regular/cyl_bessel_i0.hpp:23]: I'm all for the change, I am just not wild it being in the same diff.

File: include/eve/arch/cpu/logical_wide.hpp (hunk 2 above)
Comment (DenisYaroshevskiy) [reply include/eve/arch/cpu/logical_wide.hpp:86]: Then you should probably `requires` on them.
I don't know if you can `requires` on the nested type, if not - put it in a base class helper. (let me know if you need help to make that work).

File: include/eve/arch/cpu/wide.hpp (hunk 1 above)
Comment (DenisYaroshevskiy) [reply include/eve/arch/cpu/wide.hpp:79]: It is a public function on wide. If you don't want it to be public - make it private or put in detail.

File: include/eve/module/bessel/regular/cyl_bessel_i0.hpp (hunk 6 above)
Comment (DenisYaroshevskiy) [reply include/eve/module/bessel/regular/cyl_bessel_i0.hpp:23]: I'm not gonna make you split - but please don't do it in the future - it's extremely annoying to look through 500 files with different changes.

File: include/eve/detail/apply_over.hpp (hunk 3 above)
Comment (jfalcou) [reply include/eve/detail/apply_over.hpp:30]: I don't understand your remark.
